README.md
# SANOJ DODDAPANENI
### Project 1 - The Redactor
**Introduction** - In this project, we need to collect all the text files in a specified folder and then redact all the names, addresses, dates, genders, phone numbers. Aditionally we need to redact the synonyms for a word using concepts function. For example, we need to redact the string movies in the file when string film is mentioned while execution. This project is developed using python and command line tools in Ubuntu.  
#### Sources -   
**_For entity redaction functions -_** _https://predictivehacks.com/redact-name-entities-with-spacy/_  
**_Regex founctions and sources:_**  
_Examples -_ _https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number_  
_Tester_ - _https://www.regextester.com_  
**_File functions -_** _https://realpython.com/working-with-files-in-python/_  
**_Concepts function for synonyms -_** _https://www.geeksforgeeks.org/get-synonymsantonyms-nltk-wordnet-python/_  

**Installation directions -** In this project, we use the packages spacy, collections, re, nltk, glob, sys, os and nltk. These packages can be installed using the command below replacing [package_name] with the package that is to be installed. These packages will be imported according to requirement.  
**pipenv install [package_name]**  

Aditionally, we also need to run the command below to install required spacy model -  
```! pipenv run python -m spacy download en_core_web_sm```

#### Project Description -
**redactor.py file -** This file is provided in the project1 directory which constains the functions and execution code for the program for desired output. Each function is further explained below.

**read_text() function -** This function will read any type of the file such as text file, .md file or any other file with function readable format that is spcified in a folder and convert it into the desired format for further redaction process.

**redact_names() function -** This function takes the input from read_text() function and then using spacy package, name entities are recognized for the input and is marked with █ character.

**redact_address() function -** This function takes the input from the previous function and then using spacy module, GPE, LOC and FAC entities in the data are redacted with the █ character.

**redact_dates() function -** This function takes the input from the previous function and then using spacy, DATE entities are redated with █ character. Aditionally, regex expression for date formats -  
    ``` r'\d{4} | [\d]{1,2}/[\d]{1,2}/[\d]{2,4} | Sun,| Mon,| Tue,| Wed,| Thu,| Fri,| Sat,| Jan | Feb | Mar | Apr| May | Jun | Jul | Aug | Sep | Oct | Nov | Dec ' ```  
is used and the input file matching to the above expression is also masked with █.

**redact_genders() function -**  This function takes input from the previous called function and replaces the string with █ when matched with the format of regex expression -  ``` r'\A He | he |[Ss]he | [Hh]im| [Hh]is | [Hh]er | [Mm]en| [Ww]omen'. ```

**redact_phonenumber() function -** This function will take the input from the previous called function and then replaces the string when matched with the regex format below -  ``` r' \d{3}[-\.\s]??\d{3}[-\.\s]??\d{4}|\(\d{3}\)\s*\d{3}[-\.\s]??\d{4}|\d{3}[-\.\s]??\d{4} ' ```

**redact_concepts() function -** This function will take the input from the last called function and then using nltk package and lemmas, the line which has the word and synonym of the word mentioned during the execution of the program in command line and mask the synonyms with the character █.  
Here, while redacting the specific line containing the word or synonym of the word mentioned it will consider last \n before the word or synonym and first \n after the word and redact the line accordingly. If there are no new lines i.e., \n in the string read, it will redact the whole file as it will consider the whole file as one line without new lines. You need not specify the \n character in input text as the enter key press in input file is considered as new line while reading the text file and python replaces it with \n automatically for execution.  
In this function, splitlines() is used to convert the data to be printed in output file.

Aditionally, code for to write complete masked data is written in main function. This data will take the masked data of each  file according to the mentioned file format in a project1 folder, run the functions and then writes the redacted data into the [name].txt.redacted file in specified folder in command of project1 directory.

Further, code for stats is also written in main function, this will count the each occurance of the masked data in the input data according to each requirement names, dates, address, phone number, genders and concepts. During the execution, if we provide **stdout** or **stderr** for --stats, it will print the count of the masked data according to each file and their attributes as the terminal output. Alternatively, if we provide any other string after --stats, the count of maksed data of each file according to the file and attributes will be appended into the [name].txt file in folder stats in the directory project1 where name will be the name provided while running the command except stdout and stderr.

To run the function, we need to run the command below in project1 directory -  
```pipenv run python redactor.py --input '*.txt' --input 'project1/.txt' --names --address --dates --genders --phones --concept 'film' --output 'files/' --stats stdout```

In the command above, the word **film** can be replaced to find and mask synonym of any other word. and the word **stdout** which prints the data in console can be replaced with the word **stderr** to do the same action to print the data in the terminal or replace with any other word such as abc or xyz which will create the text file mentioned such as abc.txt and xyz.txt and will append the count of the stats in that file.

#### Test Cases - 
Here, we have created a new directory called **tests** and then created a file called **test_code.py** which contains different functions of test cases for each function in redactor.py. Each test case in the file is explained below accordingly.

Firstly, we import the packages sys to execute test file for all the directories of the project and provide the path accordingly and then import project1 folder and redactor.py within the folder and then finally we should import package pytest to run testcases accordingly. Pyest modules works on pytest framework and can be installed using the command below -  
```pipenv install pytest command.```

**test_readinput() function -** In this function, we test if there are existing files in project1 folder to be read as input. the test case will be passed if there are existing files for processing.

**test_names() function -** In this function, we declare a string and then call the function redact_names() from redactor.py and assert if the string is masked correctly. If the string declared in masked correctly, test case will be passed.

**test_address() function -** In this function, we declare a string of address and then call the function redact_address() from redactor.py and assert if the string is masked correctly. If the string declared in masked correctly, test case will be passed.

**test_dates() function -** In this function, we declare a string of address and then call the function redact_address() from redactor.py and assert if the string is masked correctly. If the string declared in masked correctly, test case will be passed.

**test_gender() function -** In this function, we declare a string from the regex expression of genders function and then call the function redact_genders() from redactor.py and assert if the string is masked correctly. If the string declared in masked correctly, test case will be passed.

**test_concept() function -** In this function, we declare a string of synonym for the word declared in variable data and then call the function redact_concepts() from redactor.py and assert if the string is masked correctly. If the string declared in masked correctly, test case will be passed.

**test_phone() function -** In this function, we declare a string of phone number and then call the function redact_phonenumber() from redactor.py and assert if the string is masked correctly. If the string declared in masked correctly, test case will be passed.

Here, Test cases can be executed using below command -  
``` pipenv run python -m pytest```

Once the command is passed, it will show the execution of test cases.

#### Possible Bugs -  
In the redactor.py program, we have used spacy module, this spacy module might not always be accurate to recognize the entities specified which could cause some data not being masked. Further, the count of concepts in the stats output might give return same values for all the text files being redacted. This bug only persists with count of stats starting from second text file.

At the end, these files are added, committed and pushed into git hub using git commands accordingly for each file.


CHANGES.md
# Changes
* **Repo not clone-able** --- I changes the repo name to the given name in the project description and the code is now clonable to run in any instance.
* **Code not executable** --- Updated the code to python 3.10.2 and added the Pipfile information with necessary modules and packages to run the code.
* **Output files not stored in respective folder** --- Made changes to the tree structure. Moved the redactor.py file and  the input files such as input.txt and input2.txt to the root directory of the project where the README.md exists for the execution. Now executing the code will redirect the output redacted files to the folder specified in the command while execution.
* **File names not re-assigned correctly** --- Made changes to the output file names from the format of [filename].txt.redacted to [filename].redacted where [filename] will take the complete name of input file including the format such input input.txt or input2.txt and will associate the .redacted format to it such as input.txt.redacted and input2.txt.redacted.
* **Missing Stats** --- Made changes to the stats file and removed the directory stats and saving the file from user input given in the command line except stdout and stderr. For example - If the command line is given as --stats statistics, it will create the statistics file with the redacted statistics written into it. If the command line is specified as --stats files/abc, it will create the abc file with the necessary information in the files directory. If stdout or stderr is given adter --stats, it will print the statistics to console and not write the statistics to any file
* **Missing/No Features Found for Dates, Concepts, Phone Number, Gender and Names** --- In this case, as the code was not executable, the functions were not working. Therefore, the changes made above will execute the code accordingly. Edited the code for genders by adding new words such as Boy, Girl, Mom, Dad etc., and addresses for more specifc redaction for full address in the form of 1800 Beaumont Drive, Norman, Oklahoma 73071.

Tested the Test cases running from the main directory to verify the successful passing of test cases.

**The command to run the code is as below** ---  
```pipenv run python redactor.py --input '*.txt' --names --address --dates --genders --phones --concept 'movie' --output 'files/' --stats stdout```

In the above command, we can replace 'film' with any other name to find synonyms and redact the sentence accordingly. Further, we can also replace files/ after output tag to any other name as name of directory to store redacted files. At the end, we can replace stdout to any other name which will create file and append the stats accordingly as mentioned in Missing Stats above.

**The command to run the test cases is as below** ---
```pipenv run python -m pytest```

**Successful execution of code** ---  
<img width="1435" alt="image" src="https://user-images.githubusercontent.com/31980486/164136335-4bcd6860-4139-44c5-884f-df035ab67812.png">  

**Successful execution of test cases** ---
<img width="1435" alt="image" src="https://user-images.githubusercontent.com/31980486/164139358-cafd86b8-0718-4e81-a7fb-783ad23f9359.png">  

**Tree Structure of the code** ---  
<img width="1435" alt="image" src="https://user-images.githubusercontent.com/31980486/164140518-ce5d28fb-9d88-479b-8b83-dc0652a877db.png">  

Changes have been made accordingly and the code has been pushed to the github and tag v2.0 is done.


tests/test_code.py
# Importing necessary modules to run test cases
# Module sys for path setting
# Importing module project1 
# Importing redactor.py from project1
# Module pytest to run test cases
# Modules nltk, spacy and os.path for entity recognition and files and directory functions

import sys
sys.path.append('..')
import redactor
import pytest
import nltk
import spacy
import en_core_web_md
import os.path

# Downloading necessary models for the pytest to run

nltk.download('wordnet')
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')
nltk.download('maxent_ne_chunker')
nltk.download('words')
nltk.download('omw-1.4')

nlp = en_core_web_md.load()

# Test case for read_text() function in redactor.py
def test_readinput():
    if os.path.exists('project1/*.txt'):
        assert True

# Test case for readact_names() function in redactor.py
def test_names():
    data = 'Mike'
    data1 = data
    data = nlp(data)
    name_test = redactor.redact_names(data1,'none')
    assert name_test == ('██████', 1)

# Test case for readct_address() function in redactor.py
def test_address():
    data = 'California'
    data1 = data
    data = nlp(data)
    address_test = redactor.redact_address(data1,'none')
    assert address_test == ('███', 1)

# Test case for readact_dates() function in redactor.py
def test_dates():
    data = ' Wed,'
    data1 = data
    data = nlp(data)
    date_test = redactor.redact_dates(data1,'none')
    assert date_test == ('█████', 1)

# Test case for readact_genders() function in redactor.py
def test_gender():
    data = ' he '
    gender_test = redactor.redact_genders(data,'none')
    assert len(gender_test) ==len('██')

# Test case for readct_concepts() function in redactor.py
def test_concept():
    data = 'film'
    concept_test = redactor.redact_concepts(data,'none','film')
    assert len(concept_test) == len('██')

# Test case for readct_phonenumber() function in redactor.py
def test_phone():
    data = '609-271-3007'
    phone_test = redactor.redact_phonenumber(data,'none')
    assert len(phone_test) ==len('██')

# Test case for writing output code in redactor.py
def test_output():
    if os.path.exists('project1/files/*.redacted'):
        assert True

# Test case for printing and writing stats of functions in redactor.py
def test_stats():
    if os.path.exists('project1/stats/*.txt'):
        assert True


redactor.py
# Importing necessary libraries to be unsed in this program.
# Module spacy for entity recognition and nlp
# Module clllections being imported
# Module re for regular expressions
# Module nltk for concepts
# Module glob for file reading
# Module sys for main function args
# Module os for file paths, files and directories

import spacy
from spacy import displacy
from collections import Counter
import re
from nltk.corpus import wordnet
import glob
import sys
import os
import nltk
import en_core_web_md


# Downloading and loading necessary models for nltk and spacy running the program

nltk.download('wordnet')
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')
nltk.download('maxent_ne_chunker')
nltk.download('words')
nltk.download('omw-1.4')

nlp = en_core_web_md.load()
#nlp = spacy.load('en_core_web_sm')

# Code for reading text file that is to be processed
def read_text(inputfiles):
    txt = []
    with open(inputfiles, 'r') as myfile:
        txt = myfile.read()
    return(txt)

# Redacting methods using spacy, re and wordnet
# Source for spacy - https://predictivehacks.com/redact-name-entities-with-spacy/
# Source for regex - https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number
# Source for regex tester - https://www.regextester.com
# Source for wordnet and lemma - https://www.geeksforgeeks.org/get-synonymsantonyms-nltk-wordnet-python/

# Redacting names using spacy entity recognition
def redact_names(txt, inputfiles):
    clean_text1 = txt
    doc = nlp(txt)
    countname1 = 0
    for ent in reversed(doc.ents):
            if ent.label_ == 'PERSON':
                clean_text1 = clean_text1[:ent.start_char] + u"\u2588"*len(ent.label_) + clean_text1[ent.end_char:]
                countname1 = countname1+1
    return clean_text1, countname1

# Redacting addresses using spacy entity recognition
def redact_address(txt, inputfiles):
    clean_text2 = txt
    doc = nlp(txt)
    countaddress1 = 0
    for ent in reversed(doc.ents):
            if ent.label_ == 'GPE':
                clean_text2 = clean_text2[:ent.start_char] + u"\u2588"*len(ent.label_) + clean_text2[ent.end_char:]
                countaddress1 = countaddress1+1
            if ent.label_ == 'LOC':
                clean_text2 = clean_text2[:ent.start_char] + u"\u2588"*len(ent.label_) + clean_text2[ent.end_char:]
                countaddress1 = countaddress1+1
            if ent.label_ == 'FAC':
                clean_text2 = clean_text2[:ent.start_char] + u"\u2588"*len(ent.label_) + clean_text2[ent.end_char:]
                countaddress1 = countaddress1+1
    str=re.findall(r"^\s*\S+(?:\s+\S+){2}\s[a-zA-Z',.\s-]{1,25}\s\d{5}(?:[-\s]\d{4})?$",txt)
    for digit in str:
        countaddress1 = countaddress1 + 1
        clean_text2=re.sub('\\b{}\\b'.format(digit),'\u2588'*len(digit), clean_text2)
    return clean_text2, countaddress1

# Redacting dates using spacy entity recognition and regex
def redact_dates(txt, inputfiles):
    clean_text3 = txt
    doc = nlp(txt)
    countdate1 = 0
    for ent in reversed(doc.ents):
            if ent.label_ == 'DATE':
                countdate1 = countdate1+1
                clean_text3 = clean_text3[:ent.start_char] + u"\u2588"*len(ent.label_) + clean_text3[ent.end_char:]
    str=re.findall(r'\d{4} | [\d]{1,2}/[\d]{1,2}/[\d]{2,4} | Sun,| Mon,| Tue,| Wed,| Thu,| Fri,| Sat,| Jan | Feb | Mar | Apr| May | Jun | Jul | Aug | Sep | Oct | Nov | Dec ',txt)
    for digit in str:
        countdate1 = countdate1+1
        clean_text3=clean_text3.replace(digit,u"\u2588"*len(digit))
    return clean_text3, countdate1

# Redacting genders using regex
def redact_genders(txt, inputfiles):
    clean_text4 = txt
    countgender1 = 0
    str=re.findall(r'(\b[Hh]usband\b|\b[Ww]ife\b|\b[Mm]other\b|\b[Ff]ather\b|\b[Mm]om\b|\b[Dd]ad\b|\b[Ss]on\b|\b[Dd]aughter|\b[Hh]imself\b|\b[Hh]erself\b|\b[Gg]irl\b|\b[Bb]oy\b|\b[Ff]emale\b|\b[Mm]ale\b|\b[Hh]e\b|\b[Ss]he\b|\b[Hh]er\b|\b[Hh]im|\b[Hh]is\b|\b[Mm]en\b|\b[Ww]omen\b|\b[Ww]oman\b|\b[Mm]an\b)',txt)
    for digit in str:
        countgender1 = countgender1 + 1
        clean_text4=re.sub('\\b{}\\b'.format(digit),'\u2588'*len(digit), clean_text4)
    return clean_text4, countgender1

# Redacting phonenumbers using regex
def redact_phonenumber(txt, inputfiles):
    clean_text5 = txt
    countphone1 = 0
    str=re.findall(r' \d{3}[-\.\s]??\d{3}[-\.\s]??\d{4}|\(\d{3}\)\s*\d{3}[-\.\s]??\d{4}|\d{3}[-\.\s]??\d{4} ',txt)
    for digit in str:
        countphone1 = countphone1+1
        clean_text5=clean_text5.replace(digit,u"\u2588"*len(digit))
    return clean_text5, countphone1

# redacting synonyms using nltk wordnet and lemma
def redact_concepts(txt, inputfiles, concept):
    clean_text6 = txt
    countconcept1 = 0
    for syn in wordnet.synsets(concept):
        for l in syn.lemmas():
            sentence = clean_text6.splitlines()
            for line in sentence:
                if concept in line:
                    countconcept1 = countconcept1 + 1
                    clean_text6=clean_text6.replace(line,u"\u2588"*len(line))
    return clean_text6, countconcept1


# Main function to run above functions
arg_ls = sys.argv
allfiles=[]
totalcount = 0
countname = 0
countaddress = 0
countdate = 0
countgender = 0
countphone = 0
countconcept = 0
for i in range(len(arg_ls)):
    if arg_ls[i] == '--input':
        # Using glob to read input files
        allfiles.extend(glob.glob(arg_ls[i+1]))

for i in allfiles:
    data = read_text(i)
    for j in range(len(arg_ls)):
            if arg_ls[j] == '--names':
                data, countname = redact_names(data, i)

            if arg_ls[j] == '--address':
                data, countaddress = redact_address(data, i)

            if arg_ls[j] == '--dates':
                data, countdate = redact_dates(data, i)

            if arg_ls[j] == '--phones':
                data, countphone = redact_phonenumber(data, i)

            if arg_ls[j] == '--concept':
                data, countconcept = redact_concepts(data, i, arg_ls[j+1])
                concept = arg_ls[j+1]

            if arg_ls[j] == '--genders':
                data, countgender = redact_genders(data, i)

            # Writing output to a file in files director
            if arg_ls[j] == '--output':
                if not os.path.exists(arg_ls[j+1]):
                    os.mkdir(arg_ls[j+1])
                z = i.split('/')[-1]
                #with open(arg_ls[j+1]+z+'.redacted', 'w') as statsfile:
                    #sys.stdout = statsfile
                file = open(arg_ls[j+1]+z+'.redacted','w')
                #sys.stdout = file
                for x in data.splitlines():
                    x = data
                file.write(data)
                file.close()

            if arg_ls[j] == '--stats':  
                # Source for file functions - https://realpython.com/working-with-files-in-python/
                # Writing stats to user dedlared file in stats folder when stderr is provided
                if arg_ls[j+1] == 'stderr':
                    print("\nRedacted Stats for input file:", i.format())
                    print("Names redacted:",countname)
                    print("Address redacted:",countaddress)
                    print("Dates redacted",countdate)
                    print("Genders redacted:",countgender)
                    print("Phonenumber redacted",countphone)
                    print("Concept Redacted",countconcept)
                
                # Printing the output of stats to console when stdout is provided
                elif arg_ls[j+1] == 'stdout':
                    print("\nRedacted Stats for input file:", i.format())
                    print("Names redacted:",countname)
                    print("Address redacted:",countaddress)
                    print("Dates redacted",countdate)
                    print("Genders redacted:",countgender)
                    print("Phonenumber redacted",countphone)
                    print("Concept Redacted",countconcept)

                else:
                    path = os.path.join('stats', arg_ls[j+1])
                    with open(arg_ls[j+1],'a') as statsfile:
                        sys.stdout = statsfile
                        print("\nRedacted Stats for input file:", i.format())
                        print("Names redacted:",countname)
                        print("Address redacted:",countaddress)
                        print("Dates redacted",countdate)
                        print("Genders redacted:",countgender)
                        print("Phonenumber redacted",countphone)
                        print("Concept Redacted",countconcept)
                    statsfile.close()
